### DISCUSSION TITLE
A for-dummies crew for trusted local (or remote) machines

### DISCUSSION THREAD

**Discussion by kendonB** [wlandau/crew#191] [Ideas]
SearchGPT estimates that there are about [30x more CPUs](https://chatgpt.com/share/673660fe-f2a4-800e-b2c4-ab53cf6ccd83) in the USA sitting inside personal machines than in cloud or HPC compute systems. At any given time, a small fraction of those personal CPUs are in use.

Could there be a crew.local package (similar to crew.cluster) which makes it easy to spin up compute on other trusted machines? I think a lot of the infrastructure is already there (see https://wlandau.github.io/crew/articles/plugins.html).

**Comment by wlandau** [comment]
Do you mean if you want workers to be on a different personal computer than the one you are running? As long as the networking allows it, a plugin could be developed. There is of course an existing `crew_controller_local()`, but all the workers are on the same machine as the local process.

**Reply by kendonB** [reply]
That's right. It would be even better if it was workers on multiple different personal computers. I'm imagining use cases like an office network where the nerds run sims with a few cores from everyone else's machines.

**Reply by shikokuchuo** [reply]
FYI mirai v2 makes this much easier. The new `ssh_config()` does the job in mirai. It would probably be easiest for a crew plugin to use that.

**Reply by wlandau** [reply]
AFAIK, SSH requires the local computer to initiate the connection and then waits for the remote machine to dial in. The plugin system in `crew` assumes the socket is already open and then workers dial in without blocking the local session. The current TLS encryption fits this model, but things like https://github.com/wlandau/crew.cluster/pull/17 are not really feasible.

**Reply by shikokuchuo** [reply]
Actually SSH logs into the remote machine (normally using pre-set credentials) by default on port 22 but could be any other, and from there it spawns a mirai daemon R session, which can either dial back directly (if the local firewall is so configured), or more likely dials into a local TCP port which is tunnelled back over the SSH connection to the same port on the local machine, where dispatcher is listening. This requires no firewall or other permission changes and is hence very user-friendly. This works not just for the local network, but anywhere you have SSH access, e.g. Amazon EC2.

**Reply by wlandau** [reply]
> Actually SSH logs into the remote machine (normally using pre-set credentials) by default on port 22 but could be any other, and from there it spawns a mirai daemon R session, which can either dial back directly (if the local firewall is so configured), or more likely dials into a local TCP port which is tunnelled back over the SSH connection to the same port on the local machine, where dispatcher is listening.

How asynchronous is this process?

**Reply by shikokuchuo** [reply]
It'd be up to the launcher I guess? In mirai I'm just using system2(wait = FALSE) to make the ssh calls async.

**Reply by shikokuchuo** [reply]
FYI SSH tunnelling will be even easier at mirai once this is merged: https://github.com/shikokuchuo/mirai/pull/209
You can see from the example syntax, that you only need to supply the IP addresses of the remote machines to SSH into.

**Reply by wlandau** [reply]
https://github.com/shikokuchuo/mirai/pull/209 might fit well with `crew`. Because it's configured with `daemons()` (plural) I would implement this at the level of [`crew_client()`](https://github.com/wlandau/crew/blob/main/R/crew_client.R) and then integrate it with all plugins. Would be really cool if all nontrivial plugins could just use SSH!

**Reply by wlandau** [reply]
On reflection, I'm not sure https://github.com/shikokuchuo/mirai/pull/209 fits because of the modifications it would require to `crew_client()`. Also, it seems like `mirai::daemons(remote = ssh_config())`  takes responsibility for worker launches, which is not really consistent with `crew`'s approach of establishing a listening socket  with `mirai::daemons()` and then allowing workers to dial in at a later point as needed.

Ideally, `mirai::daemons(url = "tcp://IP:PORT")` could be passed and IP and PORT that would work with the tunnel, then the [launcher](https://wlandau.github.io/crew/articles/plugins.html) could create the SSH tunnel if it does not already exist. If this approach is feasible, it could be its own plugin developed outside of `crew`.

Unfortunately, I don't have the means to develop it myself. My workplace prohibits ssh tunneling into our existing resources, and I can't seem to be able to create a Docker container locally that I could SSH into. https://wlandau.github.io/crew/articles/plugins.html has instructions for anyone  with a need to create their own launcher plugin.
